core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala

```diff
@@ -34,18 +37,31 @@ private[spark] class HistoryPage(parent: HistoryServer) extends WebUIPage("") {
     val requestedIncomplete =
       Option(request.getParameter("showIncomplete")).getOrElse("false").toBoolean
 
-    val allApps = parent.getApplicationList().filter(_.completed != requestedIncomplete)
-    val actualFirst = if (requestedFirst < allApps.size) requestedFirst else 0
-    val apps = allApps.slice(actualFirst, Math.min(actualFirst + pageSize, allApps.size))
-
+    val allCompletedAppsNAttempts = 
+        parent.getApplicationList().filter(_.completed != requestedIncomplete)
+    val (hasAttemptInfo, appToAttemptMap)  = getApplicationLevelList(allCompletedAppsNAttempts)
```

Comment:
We should slice on the applications, and then build the table with all the applications' attempts. So every page shows "x" applications, not attempts. After you apply the pagination logic, iterating is cheap ("x" is a small number), so calculating `hasAttemptInfo` separately is easy.

Perhaps a more efficient way of doing this would be to make `ApplicationHistoryProvider.getListing` return that map, then this code would be much simpler. It would just need to slice the map instead of transform a flat list into something different.

I just think the code you currently have is a little confusing, and a little inefficient. If you have a really long list of applications, it would waste a lot of cycles reading through it and copying things that would just be thrown away.